pkg/spoke/addon/registration_controller.go

```diff
@@ -0,0 +1,203 @@
+	addoninformerv1alpha1 "github.com/open-cluster-management/api/client/addon/informers/externalversions/addon/v1alpha1"
+	addonlisterv1alpha1 "github.com/open-cluster-management/api/client/addon/listers/addon/v1alpha1"
+)
+
+// addonRegistrationController monitors ManagedClusterAddon crs on hub and starts addon registration
+// according to the setting from cr annotations.
+type addonRegistrationController struct {
+	clusterName          string
+	agentName            string
+	baseHubKubeconfigDir string
+	addonClient          addonv1alpha1client.Interface
+	hubClientConfig      *restclient.Config
+	kubeClient           kubernetes.Interface
+	hubAddonLister       addonlisterv1alpha1.ManagedClusterAddOnLister
+	hubKubeClient        kubernetes.Interface
+	hubCSRInformer       certificatesinformers.CertificateSigningRequestInformer
+	hubCSRClient         csrclient.CertificateSigningRequestInterface
+	recorder             events.Recorder
+
+	// addonRegistrationConfigs maps the addon name to a slice of registrationConfigs
+	addonRegistrationConfigs map[string][]registrationConfig
+	// addonCertificateManagers maps the hash of a registrationConfig to a addonCertificateManager
+	addonCertificateManagers map[string]*addonCertificateManager
+}
+
+// NewAddonRegistrationController returns an instance of addonRegistrationController
+func NewAddonRegistrationController(
+	clusterName string,
+	agentName string,
+	baseHubKubeconfigDir string,
+	kubeClient kubernetes.Interface,
+	hubClientConfig *restclient.Config,
+	hubKubeClient kubernetes.Interface,
+	hubCSRInformer certificatesinformers.CertificateSigningRequestInformer,
+	hubAddonInformers addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	hubCSRClient csrclient.CertificateSigningRequestInterface,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &addonRegistrationController{
+		clusterName:          clusterName,
+		agentName:            agentName,
+		baseHubKubeconfigDir: baseHubKubeconfigDir,
+		kubeClient:           kubeClient,
+		hubClientConfig:      hubClientConfig,
+		hubKubeClient:        hubKubeClient,
+		hubAddonLister:       hubAddonInformers.Lister(),
+
+		hubCSRInformer: hubCSRInformer,
+		hubCSRClient:   hubCSRClient,
+
+		recorder: recorder,
+
+		addonRegistrationConfigs: map[string][]registrationConfig{},
+		addonCertificateManagers: map[string]*addonCertificateManager{},
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				accessor, _ := meta.Accessor(obj)
+				return accessor.GetName()
+			},
+			hubAddonInformers.Informer()).
+		WithSync(c.sync).
+		ToController("AddonClientCertManagerController", recorder)
+}
+
+func (c *addonRegistrationController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	addonName := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling addon %q", addonName)
+
+	cachedConfigs := c.addonRegistrationConfigs[addonName]
```

Comment:
do we need to keep the cached configs for cleanup? We may not need to handle that because cert will expire and hub controller to setup rbac to disallow the cert. 

The  problem here is if this is restarted, we still cannot guarantee the cleanup.